processing/src/main/java/io/druid/query/aggregation/cardinality/CardinalityAggregator.java

```diff
@@ -23,77 +23,53 @@
 import com.google.common.hash.Hasher;
 import com.google.common.hash.Hashing;
 import io.druid.query.aggregation.Aggregator;
+import io.druid.query.QueryDimensionInfo;
 import io.druid.query.aggregation.hyperloglog.HyperLogLogCollector;
-import io.druid.segment.DimensionSelector;
-import io.druid.segment.data.IndexedInts;
 
-import java.util.Arrays;
 import java.util.List;
 
 public class CardinalityAggregator implements Aggregator
 {
-  private static final String NULL_STRING = "\u0000";
+  public static final String NULL_STRING = "\u0000";
 
   private final String name;
-  private final List<DimensionSelector> selectorList;
+  private final List<QueryDimensionInfo> dimInfoList;
   private final boolean byRow;
 
-  private static final HashFunction hashFn = Hashing.murmur3_128();
+  public static final HashFunction hashFn = Hashing.murmur3_128();
   public static final char SEPARATOR = '\u0001';
 
-  protected static void hashRow(List<DimensionSelector> selectorList, HyperLogLogCollector collector)
+  protected static void hashRow(List<QueryDimensionInfo> dimInfoList, HyperLogLogCollector collector)
   {
     final Hasher hasher = hashFn.newHasher();
-    for (int k = 0; k < selectorList.size(); ++k) {
+    for (int k = 0; k < dimInfoList.size(); ++k) {
       if (k != 0) {
         hasher.putByte((byte) 0);
       }
-      final DimensionSelector selector = selectorList.get(k);
-      final IndexedInts row = selector.getRow();
-      final int size = row.size();
-      // nothing to add to hasher if size == 0, only handle size == 1 and size != 0 cases.
-      if (size == 1) {
-        final String value = selector.lookupName(row.get(0));
-        hasher.putUnencodedChars(value != null ? value : NULL_STRING);
-      } else if (size != 0) {
-        final String[] values = new String[size];
-        for (int i = 0; i < size; ++i) {
-          final String value = selector.lookupName(row.get(i));
-          values[i] = value != null ? value : NULL_STRING;
-        }
-        // Values need to be sorted to ensure consistent multi-value ordering across different segments
-        Arrays.sort(values);
-        for (int i = 0; i < size; ++i) {
-          if (i != 0) {
-            hasher.putChar(SEPARATOR);
-          }
-          hasher.putUnencodedChars(values[i]);
-        }
-      }
+
+      QueryDimensionInfo dimInfo = dimInfoList.get(k);
+      dimInfo.queryHelper.hashRow(dimInfo.selector, hasher);
```

Comment:
I'm leaning towards keeping dimInfo as is, I wanted that to be strictly a "bundle of information" and all of the query processing functionality to reside in the DimensionQueryHelper interface